### PR TITLE
migration: bucket and handler

### PR DESCRIPTION
**Pull request by husio**
A bucket and message handler implementation added.

Writing more code allowed me to see the use case more clearly. I have to make some design decisions before continuing.
Below are some of the thought and open questions. Please let me know what you think on each topic.

###  `Copy() orm.CloneableData` must clone the metadata as well
The `Copy` method must [clone the new field as well](https://github.com/iov-one/weave/compare/schema_migrations_issue_537...schema_migrations_issue_537_handlers?expand=1#diff-6bdbeffb7a0fcce672abbcd92dedf1c3R29). I think it is important to somehow enforce that the copy of `metadata` attribute is made and not forgotten during migration or a new model implementation.

### `init` function is required
I think it is required to use `init` function to [register the migration function](https://github.com/iov-one/weave/compare/schema_migrations_issue_537...schema_migrations_issue_537_handlers?expand=1#diff-6bdbeffb7a0fcce672abbcd92dedf1c3R13). I don't think there is any better way to do this.

### Metadata should be part of the `orm`
I think there is only one way a metadata can be transparently enforced and validated. It must be part of the `orm` package. Metadata must be [validated shortly before save](https://github.com/iov-one/weave/compare/schema_migrations_issue_537...schema_migrations_issue_537_handlers?expand=1#diff-00477cb27bcef9d4d5b11c0ac3479c22R188) and model must be [migrated to the latest version before returning](https://github.com/iov-one/weave/compare/schema_migrations_issue_537...schema_migrations_issue_537_handlers?expand=1#diff-00477cb27bcef9d4d5b11c0ac3479c22R166)
This makes metadata and schema versioning not optional but required.

### Schema versioning must be available for messages
Metadata cannot be part of the `orm` because it must be available for the `app` package as well. Router must [upgrade the message](https://github.com/iov-one/weave/compare/schema_migrations_issue_537...schema_migrations_issue_537_handlers?expand=1#diff-a6456faf55d0ef22a0827751abbd2e55R72) before `Check`/`Deliver` call.

### CODE REVIEW

**Comment by ethanfrey** [comment]
First comments on the description, then code review:

> I think it is required to use init function to register the migration function. I don't think there is any better way to do this.

Yes, I approve of this. Each module in msg.go/models.go can register migrations next to the data structures. Seems safest to me.

> Metadata should be part of the orm

Yes, we are making a decision of the "right way" to version and upgrade and add deep support for it. You can theoretically build an app or modules without using Buckets (app package only depends on the implementation of store objects). Let's make the orm package opinionated, and easy to use if you follow our (to be documented) "best practices". However, orm should *not* depend on many other packages... so let's check the imports.

> Schema versioning must be available for messages

Ha, now you got me in a trap :)

I would say the following:
* Metadata protoc and interface defined in `weave` (top-level) along with all other types/interfaces used throughout the project
* Migrations package has no dependencies except `weave` and `errors` and is intended to be imported by other packages (we should document this... which packages are first-level and can be imported by siblings, to avoid circular imports)
* app package can import `weave` and `migrations` to perform checks/migrations on imcoming messages.
* orm package can import `weave` and `migrations` to perform checks/migrations on saving models
* migrations package must import `orm` to store internal state in a Bucket.

oops... circular import coming (orm <-> migrations)

What about orm not knowing about migrations, but migrations knowing about orm. And it can expose a "MigrationsAwareBucket" (please find a better name) that wraps a orm.Bucket and handles the auto-migration logic. Also a migrations.Object that does better than the current orm.Object for this Metadata.

And then... the more I think of it, if we have app import migrations that imports orm and every extension will also import migrations to import orm... why not just move the contents of migration into the orm package? It is already imported by all `x/*` modules, no big deal to import it from `app` which is also quite high-level. (We really need to map out the dependency graph soon)

File: migration/handler.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func (h *upgradeSchemaHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	ver, err := h.bucket.CurrentSchema(db, msg.Pkg)
+	if err != nil && !errors.ErrNotFound.Is(err) {
```
Comment (ethanfrey) [inline migration/handler.go]: Nice, so this same message will upgrade form 0 to 1 (ErrNotFound) to enable it in the first place. I like the symetry

File: migration/handler.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	schema := Schema{Pkg: msg.Pkg, Version: ver + 1}
+	obj, err := h.bucket.Create(db, &schema)
+	if err != nil {
+		return nil, errors.Wrap(err, "create schema version")
+	}
+
+	return &weave.DeliverResult{Data: obj.Key()}, nil
+}
+
+func (h *upgradeSchemaHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpgradeSchemaMsg, error) {
```
Comment (ethanfrey) [inline migration/handler.go]: Nice. The handler code is looking shorter and shorter, less boilerplate.

However, we really do need some permission here, who can do the upgrade. Can you path an Address in the constructor of the Handler and enforce that here? It can be a more generic `gconf` solution later, but at least a placeholder for such authorization logic.

File: migration/model.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		Version:  s.Version,
+		Pkg:      s.Pkg,
+	}
+}
+
+// schemaID returns a detereministic ID of this schema instance.
+func schemaID(pkg string, version uint32) []byte {
+	raw := make([]byte, len(pkg)+4)
+	copy(raw, pkg)
+	binary.LittleEndian.PutUint32(raw[len(pkg):], version)
```
Comment (ethanfrey) [inline migration/model.go]: Why LittleEndian?
We use BigEndian everywhere else for the reason that it provides lexographical ordering.

Eg. If I do a range scan for `cash` o `casi`, I will get `cash:0x1`, `cash:0x2`, `cash:0x3` in that order. If I use LittleEndian, I have no guarantees... `cash:0x0100`, `cash:0x0001`, `cash:0x0101`

Please use BigEndian for any integers that will be used in a key (and in general for consistency) - 6 times in our code vs. 0 for LittleEndian

Comment (ethanfrey) [reply migration/model.go]: Protobuf varint is kind of like a compressed little-endian, but we don't use that for any keys

File: migration/model.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+func NewSchemaBucket() *SchemaBucket {
+	b := orm.NewBucket("schema", orm.NewSimpleObj(nil, &Schema{}))
+	return &SchemaBucket{Bucket: b}
+}
+
+// CurrentSchema returns the current version of the schema for a given package.
+// It returns ErrNotFound if no schema version was registered for this package.
+// Minimum schema version is 1.
+func (b *SchemaBucket) CurrentSchema(db weave.KVStore, packageName string) (uint32, error) {
+	for ver := uint32(1); ver < 10000; ver++ {
```
Comment (ethanfrey) [inline migration/model.go]: If you use BigEndian (above), you can use a reverse iterator as I mentioned, and grab the first element from it. This is one read and will not have a limit.

File: migration/model.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// Save persists the state of a given schema entity.
+func (b *SchemaBucket) Save(db weave.KVStore, obj orm.Object) error {
+	s, ok := obj.Value().(*Schema)
+	if !ok {
+		return errors.Wrapf(errors.ErrInvalidModel, "invalid type: %T", obj.Value())
+	}
+	switch ver, err := b.CurrentSchema(db, s.Pkg); {
+	case err != nil:
```
Comment (ethanfrey) [inline migration/model.go]: Don't we need to explicitly handle errors.ErrNotFound here to allow the first save?

Comment (ethanfrey) [reply migration/model.go]: I also don't see any readability improvements here of this switch statement over two if statements checking for error and invalid version upgrade. In fact the opposite.

File: migration/model.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	case ver+1 != s.Version:
+		// Schema versioning is sequential and the numbers must be incrementing.
+		return errors.Wrapf(errors.ErrInvalidInput, "previous schema is %d", ver)
+	}
+	return b.Bucket.Save(db, obj)
+}
+
+// Create adds given schema instance to the store and returns the ID of the
+// newly inserted entity.
+func (b *SchemaBucket) Create(db weave.KVStore, s *Schema) (orm.Object, error) {
```
Comment (ethanfrey) [inline migration/model.go]: Why is create like save?
I assume it would require b.CurrentSchema to be `ErrNotFound` and then start at any number (so we can initialize anywhere from the genesis file. I am a bit confused where the two should be used and why they are the same

File: orm/bucket.go [hunk 7]
```diff
@@ -182,6 +185,8 @@ func (b Bucket) Save(db weave.KVStore, model Object) error {
 		return err
 	}
 
+	// TODO - ensure the metadata is set
```
Comment (ethanfrey) [inline orm/bucket.go]: We also want to ensure latest version before Save (and error otherwise, or auto-migrate?)

File: orm/bucket.go [hunk 8]
```diff
@@ -162,6 +162,9 @@ func (b Bucket) Parse(key, value []byte) (Object, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO - ensure the value is migrated to the latest version
```
Comment (ethanfrey) [inline orm/bucket.go]: Yes, automigration on read is good

File: orm/bucket.go (hunk 7 above)
Comment (ethanfrey) [reply orm/bucket.go]: Of course, knowing what the latest version is, and how to migrate means importing `migrations`, which in turn imports `orm` for Bucket.

Let's just merge those two packages. Unless you have a better idea

File: migration/handler.go (hunk 2 above)
Comment (husio) [reply migration/handler.go]: Done in https://github.com/iov-one/weave/pull/550/commits/bf6175479febe71e130d6a12be63383d2efa2d18

File: migration/model.go (hunk 3 above)
Comment (husio) [reply migration/model.go]: Usually I follow [a rule](https://twitter.com/rob_pike/status/819768645182103553) when deciding this. It makes sense to change to BigEndian.

Comment (husio) [reply migration/model.go]: Done in https://github.com/iov-one/weave/pull/550/commits/f5511e5e0eefb19afb4101f52bb2fed2784df3f5

File: migration/model.go (hunk 4 above)
Comment (husio) [reply migration/model.go]: I was thinking about it. I will leave it for now and come back to this before merging to master.

File: migration/model.go (hunk 5 above)
Comment (husio) [reply migration/model.go]: Done in https://github.com/iov-one/weave/pull/550/commits/dd832b554723c1de1ebae6d444dc1f65863fc7b5

File: migration/model.go (hunk 6 above)
Comment (husio) [reply migration/model.go]: I was following [`Create`](https://github.com/iov-one/weave/blob/master/x/distribution/model.go#L111) and [`Save`](https://github.com/iov-one/weave/blob/master/x/distribution/model.go#L131) implementations from other extensions.
I agree the naming doesn't make the functionality obvious. We tried to address this issue at least once already :(

File: orm/bucket.go (hunk 7 above)
Comment (husio) [reply orm/bucket.go]: I am all for merging. Until I start working on the messages which require this functionality as well. [Transaction](https://github.com/iov-one/weave/pull/550#discussion_r278047980) might need it too. :see_no_evil:

File: migration/configuration.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package migration
+
+import (
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+)
+
+func init() {
```
Comment (ethanfrey) [inline migration/configuration.go]: Great, we can unit test migrations with itself :)

Comment (husio) [reply migration/configuration.go]: We are [almost there](https://www.youtube.com/watch?v=rc7-HbYvD9I)!

**Comment by husio** [comment]
I have continued on the topic of [where to put the schema migration code](https://github.com/iov-one/weave/pull/550#issuecomment-486148577) and I do not have a solution yet. Just a few findings that I will try to summarize below.

### The requirements

1. We want to avoid the need for manual migration. This means that
 a. when fetching a model from a bucket, I want it to be automagically migrated if needed
 b. before storing a model in a bucket, I want to make sure it is migrated to the latest version
 c. when processing a transaction in a handler, I want the message to be migrated to the latest schema version
2. Schema migration code is registered in Go, but the switch to the N version is done via message, that is processed by a handler
 a. Only an admin must be able to upgrade the schema. This means that a configuration is required so that we can load from genesis (and update if needed) the admin address, 
 b. Each package has a separate counter for the schema number

### The issues
- [ ]  including desired migration functionality (1.a) inside of a bucket is right now not possible because:
 - bucket `Save` and `Get` methods do not know which package they belong to. This is required by 2.b
 - bucket `Save` and `Get` method do not have access to the context which is required by the migration function
- [ ] when `app.Router` is migrating a message, it does not know which package it comes from so it cannot get the schema version for the package (2.b)
- [ ] right now schema migration is in `migration` package. This should be moved to `weave` as both `weave/orm` and `weave/app` packages require it. While this will compile, `weave` package will contain components of an extension - a model, a handler, an initializer, a message declaration, configuration etc. Those compontents should not be part of the top level `weave` package

My progress towards a working code is in https://github.com/iov-one/weave/commit/989109b60244342088569971c8bea72bc3358bf0
Main functionality (incomplete without an implementation!) is in `weave` package ([`schema.go`](https://github.com/iov-one/weave/commit/989109b60244342088569971c8bea72bc3358bf0#diff-9fc4fe326c976ef4c2c6838a096e415d))

Any ideas how to solve any of those issues? The easiest solution is to give up on automatic migration and require the user to explicitely call `migration.Apply(... &mydata)`, but that would sux.

**Comment by ethanfrey** [comment]
Some thoughts here... maybe good, maybe bad... but hopefully they will combine with yours to help find a good solution to the structure.

### Opinions

1. I don't really like auto-magic names, as it does save a little bit of typing but makes it hard to follow the code. If this is the only clean solution, however, you can use them.
2. I think we can separate Model and Msg migration logic. We may be able to customize the various migrators to make this easier.


### Ideas

> bucket Save and Get methods do not know which package they belong to...

1. If we use batch updates, I can view a way of handling the bucket updates.
  - Define `type ModelMigrator func(db weave.KVStore) error` which will update all models in the key store. It will be defined in and registered from the proper package and know all the buckets and such. This is much more like how we handle db migrations from webapps, and I am okay for a few seconds extra block time on upgrades (still better than a few hours downtime)
  - We now know that all data read from the Bucket will be the proper version
  - We _assume_ the code called when msg has schema N will store data in schema N. This is an issue that should be covered in tests. While it is nice to provide a run-time guard, we don't need to if this is a blocker.

> when app.Router is migrating a message, it does not know which package it comes

2. This one is tricky. But we can also use a naming standard (well documented) instead of magic. For example, all messages in cash [have a path that starts with `cash/`](https://github.com/iov-one/weave/blob/dd832b554723c1de1ebae6d444dc1f65863fc7b5/x/cash/msg.go#L13-L15). Same with [escrow](https://github.com/iov-one/weave/blob/dd832b554723c1de1ebae6d444dc1f65863fc7b5/x/escrow/msg.go#L9-L14) and I believe all other packages.

I would parse out the name before the slash in the Path as the package name. We need to make sure we use the same one to register the model migrations. With that assumption (well documented), this becomes easier.

>  right now schema migration is in migration package.

3. We also import `errors` and `crypto` and `weavetest` from lots of places. We also import `orm` from `x/...` Let's just import `migrations` from `orm` and `app`. I think it is not dangerous. Top-level would be horrible, as it would import `orm` from `weave` and `weave` from `orm`. Just leave it there for now, and we can figure out another place later if needed, but I think it is the best we can do.

**Comment by husio** [comment]
I have implemented something that I think is a good solution.

All code is in `migration` package.

When you want your model schema to be version controlled, you must use `migration.SchemaVersionedBucket` instead of `orm.Bucket`. It has the same interface but does the extra migration work. An example can be found in the test file https://github.com/iov-one/weave/blob/c4117093fbbdd9e1207ed2d66b2db1f5eb9a6c9a/migration/orm_test.go

When you want your messages schema to be version controlled, you must wrap the handler with `migration.SchemaMigratingHandler`. It returns a `weave.Handler` that does the extra migration work. An example code can be found in the test file https://github.com/iov-one/weave/blob/c4117093fbbdd9e1207ed2d66b2db1f5eb9a6c9a/migration/handler_test.go

@ethanfrey please have a look and let me know what you think.

**Comment by ethanfrey** [comment]
I like that idea.

So neither orm nor app knows about migrations, but rather the extensions themselves pull it in.
Very nice and modules.

**Comment by ethanfrey** [comment]
Looking at code now

File: migration/orm_test.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		Cnt:      17,
+	})
+	assert.Nil(t, b.Save(db, obj))
+}
+
+type MyModelBucket struct {
+	Bucket
+}
+
+func (b *MyModelBucket) GetMyModel(db weave.KVStore, key string) (*MyModel, error) {
```
Comment (ethanfrey) [inline migration/orm_test.go]: We really need to get rid of this boilerplate one day...

File: migration/orm_test.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+		msg.Cnt += 2
+		return msg.err
+	})
+	reg.MustRegister(3, &MyModel{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
+		panic("not implemented")
+	})
+
+	db := store.MemStore()
+
+	schema := NewSchemaBucket()
```
Comment (ethanfrey) [inline migration/orm_test.go]: okay, so this is the global bucket that stores all schema info

File: migration/orm_test.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	if _, err := schema.Create(db, &Schema{Metadata: &weave.Metadata{Schema: 1}, Pkg: thisPkgName, Version: 1}); err != nil {
+		t.Fatalf("cannot register schema version: %s", err)
+	}
+
+	b := &MyModelBucket{
+		Bucket: NewBucket(thisPkgName, "mymodel", orm.NewSimpleObj(nil, &MyModel{})),
```
Comment (ethanfrey) [inline migration/orm_test.go]: And this uses `migration.Bucket` instead of `orm.Bucket`, right?
With just one more argument so it is package-schema aware

File: migration/orm.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+	schema      *SchemaBucket
+	migrations  *register
+}
+
+func NewBucket(packageName string, bucketName string, model orm.Cloneable) Bucket {
+	return Bucket{
+		Bucket:      orm.NewBucket(bucketName, model),
+		packageName: packageName,
+		schema:      NewSchemaBucket(),
+		migrations:  reg,
```
Comment (ethanfrey) [inline migration/orm.go]: Ah, so you copy a reference here to the global register with the purpose of overriding it for test cases, right? There is no other need to copy this reference.

But it does seem like a good way to handle mocking. Maybe add a function to set it, so tests read clearer.

`func (svb *Bucket) mockMigrations()` -> `svb.migrations = newRegistry()`

And then just register migrations on the bucket with a public helper (to global namespace unless mocked for tests):

`func (svb *Bucket) MustRegister(...)` -> `svb.migrations.MustRegister(...)`

One can also use this in the init functions. Maybe even pulling the model automatically from the bucket (in a future PR)

File: migration/orm_test.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+	})
+	assert.Nil(t, b.Save(db, obj))
+
+	if m, err := b.GetMyModel(db, "schema_one"); err != nil {
+		t.Fatalf("cannot get model one: %s", err)
+	} else if m.Metadata.Schema != 1 || m.Cnt != 5 {
+		t.Fatalf("unexpected result model: %#v", m)
+	}
+
+	// Storing a model with a schema version higher than currently active
```
Comment (ethanfrey) [inline migration/orm_test.go]: :+1:

File: migration/orm_test.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+	// is not allowed.
+	obj = orm.NewSimpleObj([]byte("schema_two"), &MyModel{
```
Comment (ethanfrey) [inline migration/orm_test.go]: I would use obj2 here for more clarity. I was confused, thinking you saved the same object again. When you did save a new object.

File: migration/orm_test.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+		t.Fatalf("cannot save second object after schema version update: %s", err)
+	}
+
+	// Now that the schema was upgraded, all returned modlels must use it.
+	// This means that returned models metadata schema must be set to two
+	// and the payload must be updated.
+
+	if m, err := b.GetMyModel(db, "schema_one"); err != nil {
+		t.Fatalf("cannot get first model: %s", err)
+	} else if m.Metadata.Schema != 2 || m.Cnt != 5+2 {
```
Comment (ethanfrey) [inline migration/orm_test.go]: Very nice demo of migrations

File: migration/orm_test.go [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+	reg := newRegister()
+
+	reg.MustRegister(1, &MyModel{}, NoModification)
+	reg.MustRegister(2, &MyModel{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
+		msg := m.(*MyModel)
+		msg.Cnt += 2
+		return msg.err
+	})
+	reg.MustRegister(3, &MyModel{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
```
Comment (ethanfrey) [inline migration/orm_test.go]: Why do you register this?

Either test it, or remove this line

File: migration/handler_test.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+	schema := NewSchemaBucket()
+	if _, err := schema.Create(db, &Schema{Metadata: &weave.Metadata{Schema: 1}, Pkg: thisPkgName, Version: 1}); err != nil {
+		t.Fatalf("cannot register schema version: %s", err)
+	}
+
+	handler := SchemaMigratingHandler(thisPkgName, &weavetest.Handler{})
+	// Use custom register reference so that our test is not polluted by
+	// extenrnal registrations.
+	handler.(*schemaMigratingHandler).migrations = reg
```
Comment (ethanfrey) [inline migration/handler_test.go]: I would suggest using some cleaner mock setup like I proposed for Bucket.
If `SchemaMigratingHandler` returns `*schemaMigratingHandler` not the interface, then you have access to those extra methods, at least in test code.

File: migration/handler_test.go [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/iov-one/weave/weavetest/assert"
+)
+
+func TestSchemaMigratingHandler(t *testing.T) {
+	const thisPkgName = "testpkg"
+
+	reg := newRegister()
+
+	reg.MustRegister(1, &MyMsg{}, NoModification)
+	reg.MustRegister(2, &MyMsg{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
```
Comment (ethanfrey) [inline migration/handler_test.go]: Ahh... so you edit the message in place.
Hmmm... Somehow I prefer passing a new one, but the truth is this makes it much cleaner to reuse the same transaction.

File: migration/handler_test.go [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+		msg := m.(*MyMsg)
+		msg.Content += " m2"
+		return msg.err
+	})
+	reg.MustRegister(3, &MyMsg{}, func(db weave.ReadOnlyKVStore, m Migratable) error {
```
Comment (ethanfrey) [inline migration/handler_test.go]: Again, why register this if not tested

File: migration/handler_test.go [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+	// extenrnal registrations.
+	handler.(*schemaMigratingHandler).migrations = reg
+
+	var err error
+
+	// Message has the same schema version as currently active one. No
+	// migration should be applied.
+	// Handler is modyfing/migrating message in place so we can use `msg`
+	// reference to check migrated message state.
+	msg := &MyMsg{
```
Comment (ethanfrey) [inline migration/handler_test.go]: Good set of tests below.
The look ripe for migrating to table-driven tests

File: migration/handler_test.go (hunk 19 above)
Comment (husio) [reply migration/handler_test.go]: I is not possible to implement it with a copy making. There is `Copy()` method, but because of types working with such code would be very hard. It would be also hard to update the references to message/model from old to new one (ie for transaction).

File: migration/orm_test.go (hunk 11 above)
Comment (husio) [reply migration/orm_test.go]: Yes.
It is a bucket like any others. It holds `Schema` entries and no special logic. It is a basic key-value store for package -> schema version storage.

Register on the other hand does not know anything about package versions. It can apply migration only if you tell it how far to migrate https://github.com/iov-one/weave/blob/c4117093fbbdd9e1207ed2d66b2db1f5eb9a6c9a/migration/register.go#L94

Those two are merged in custom bucket and handler implementations - there the schema version for the package is taken out of the database and register is told to apply migrations up to current schema version  https://github.com/iov-one/weave/blob/c4117093fbbdd9e1207ed2d66b2db1f5eb9a6c9a/migration/orm.go#L48-L64

File: migration/orm_test.go (hunk 10 above)
Comment (husio) [reply migration/orm_test.go]: This is not hard to implement. It is problematic to change the API because it will require a lot of code to be changed. :see_no_evil:

File: migration/orm.go (hunk 13 above)
Comment (husio) [reply migration/orm.go]: Done in https://github.com/iov-one/weave/pull/550/commits/88270383a81e3290ce2993b3fea9beaa8579bfc5

File: migration/handler_test.go (hunk 18 above)
Comment (husio) [reply migration/handler_test.go]: I did not want to expose this handler implementations, that's why I'm returning `weave.Handler`. Making this a public method would require `register` to be public as well. 

Instead I have created a private function to set this.  https://github.com/iov-one/weave/pull/550/commits/88270383a81e3290ce2993b3fea9beaa8579bfc5#diff-1c3925c3634d36476b64497288c1c6b3R122
